xgi/generators/random.py

```diff
@@ -1,45 +1,129 @@
 """Generate random (non-uniform) hypergraphs."""
 
-import math
 import random
 import warnings
 from collections import defaultdict
 from itertools import combinations
+from math import floor, log
+from warnings import warn
 
 import numpy as np
 from scipy.special import comb
 
 from .classic import empty_hypergraph
 from .lattice import ring_lattice
+from .uniform import _index_to_edge_comb
 
 __all__ = [
+    "fast_random_hypergraph",
     "random_hypergraph",
     "chung_lu_hypergraph",
     "dcsbm_hypergraph",
     "watts_strogatz_hypergraph",
 ]
 
 
-def random_hypergraph(N, ps, order=None, seed=None):
+def fast_random_hypergraph(n, ps, order=None, seed=None):
+    """Generates a random hypergraph with a fast algorithm.
+
+    Generate `n` nodes, and connect any `d+1` nodes
+    by a hyperedge with probability `ps[d-1]`.
+
+    This uses a fast method for generating hyperedges.
+    See the references for more details.
+
+    Parameters
+    ----------
+    n : int
+        Number of nodes
+    ps : list of float, or float
+        List of probabilities (between 0 and 1) to create a
+        hyperedge at each order d between any d+1 nodes. For example,
+        ps[0] is the wiring probability of any edge (2 nodes), ps[1]
+        of any triangles (3 nodes). If a float, generate a uniform hypergraph
+    order: int, list of ints, or array of ints or None (default)
+        If None, ignore. If list or array, generates a hypergraph
+        with edges of orders `order[0]`, `order[1]`, etc.
+        (The length of `ps` must match the length of `order` in this case).
+    seed : integer or None (default)
+            Seed for the random number generator.
+
+    Returns
+    -------
+    Hypergraph object
+        The generated hypergraph
+
+    References
+    ----------
+    M. Dewar et al.
+    "Subhypergraphs in non-uniform random hypergraphs"
+    https://arxiv.org/abs/1703.07686
+
+    Nicholas W. Landry and Juan G. Restrepo,
+    "Opinion disparity in hypergraphs with community structure",
+    Phys. Rev. E **108**, 034311 (2024).
+    https://doi.org/10.1103/PhysRevE.108.034311
+
+    See Also
+    --------
+    random_hypergraph
+
+    Example
+    -------
+    >>> import xgi
+    >>> H = xgi.fast_random_hypergraph(50, [0.1, 0.01])
+    """
+    if seed is not None:
+        random.seed(seed)
+
+    ps, order = _check_input_args(ps, order)
+
+    nodes = range(n)
+
+    H = empty_hypergraph()
+    H.add_nodes_from(nodes)
+
+    for d, p in zip(order, ps):
+        if p == 1.0:
```

Comment:
```suggestion
        if p == 1:
```
don't need the .0 anymore i think